### DISCUSSION TITLE
Best way to validate and reset writable atom (specifically atomWithHash)

### DISCUSSION THREAD

**Discussion by mwisnicki** [pmndrs/jotai#1849] [Q&A]
Let's say I have some writable atom such as `baseAtom = atomWithHash<T>("p","")`.
Additionally I have validation function stored in another atom that always returns valid value (e.g. a default in case of of input) `validatorAtom = Atom<{(x:T) =>T}>`

What would be the best way to prevent `baseAtom` from having invalid value?

I know how to get validated value:

```ts
const validatedAtom = atom(get => get(validatorAtom)(get(baseAtom))));
```

but this won't reset original atom, so for example in the case of atomWithHash my hash parameter would be wrong.

Currently I use a hook but would prefer to do this completely within jotai.

```ts
function useValidateParams() {
  const [p, setP] = useAtom(baseAtom);
  const validator = useAtomValue(validatorAtom);
  useEffect(() => {
    setP(validator(p));
  }, [p]);
}
```

**Comment by dai-shi** [comment]
By resetting the original atom, do you mean to change the url hash too?
useEffect is a valid solution, or if you want to stay in Jotai world, store api might help (hmm, but it still requires useEffect unless you have your own single store.)

But, what's ideal is to create your own atomWithHash to have more features.
For now, you can customize `deserialize` which is the place you can validate (as its intended usage) unless you need to change the url hash.

**Reply by mwisnicki** [reply]
You're right. Looks like I'd need to fork atomWithHash to get desired functionality.
Or make a feature request 😉
